File: src/main/java/com/hp/jipp/encoding/IntegerType.java
package com.hp.jipp.encoding;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;

public class IntegerType extends AttributeType<Integer> {
    private static final String TYPE_NAME = "Integer";

    static final Attribute.SimpleEncoder<Integer> ENCODER = new Attribute.SimpleEncoder<Integer>(TYPE_NAME) {
        @Override
        public Integer readValue(DataInputStream in, Tag valueTag) throws IOException {
            Attribute.expectLength(in, 4);
            return in.readInt();
        }

        @Override
        public void writeValue(DataOutputStream out, Integer value) throws IOException {
            out.writeShort(4);
            out.writeInt(value);
        }

        @Override
        public boolean valid(Tag valueTag) {
            return Tag.IntegerValue.equals(valueTag) || Tag.EnumValue.equals(valueTag);
        }
    };

    public IntegerType(Tag tag, String name) {
        super(ENCODER, tag, name);
    }
}

File: src/main/java/com/hp/jipp/encoding/IntegerType.kt
// Copyright 2017 HP Development Company, L.P.
// SPDX-License-Identifier: MIT

package com.hp.jipp.encoding

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException

/** Attribute type for [Int] attributes. May also encode an enum type if it is not recognized. */
open class IntegerType(tag: Tag, override val name: String) : AttributeType<Int>(Encoder, tag) {

    constructor(name: String) : this(Tag.integerValue, name)

    companion object Encoder : SimpleEncoder<Int>("Integer") {
        @Throws(IOException::class)
        override fun readValue(input: DataInputStream, valueTag: Tag): Int {
            input.takeLength(INT_LEN)
            return input.readInt()
        }

        @Throws(IOException::class)
        override fun writeValue(out: DataOutputStream, value: Int) {
            out.writeShort(INT_LEN)
            out.writeInt(value)
        }

        override fun valid(valueTag: Tag): Boolean {
            return Tag.integerValue == valueTag || Tag.enumValue == valueTag
        }
    }
}
